### ISSUE TITLE
License Enforcement

### ISSUE THREAD

**Issue by afroDC** [GluuFederation/cluster-mgr#83]
Add license enforcement to cluster manager. The application needs to check against a license server to ensure the user has a license.

**Comment by afroDC** [comment]
According to Gerald, the license server has an API we can call to verify a license. Alternatively I would like to have some sort of offline solution that can detect if a license is valid as well.

**Comment by iromli** [comment]
There's `oxlicense-validator` (a JAR file) which able to _decodes_ signed license retrived from license server.

**Comment by afroDC** [comment]
@iromli so if a customer goes through the process of getting the license through our marketplace, we could use that license with the oxlicense-validator on a non-internet facing computer and unlock the program?

**Comment by yuriyz** [comment]
For license generation and validation it's required to have:
- license_id
- public_key
- public_password 
- license_password

It is trivial task if java is used. We have oxlicense-client (to fetch license, send statistic and others) and oxlicense-validator (to validate license)

Please find maven artifacts here
```
https://ox.gluu.org/maven/org/xdi/oxlicense-client/3.2.0-SNAPSHOT/
https://ox.gluu.org/maven/org/xdi/oxlicense-validator/3.2.0-SNAPSHOT/
```

See example on how to validate license here:
```
https://github.com/GluuFederation/oxd/blob/version_3.1.1/oxd-server/src/main/java/org/xdi/oxd/server/license/LicenseService.java#L108-108
```
To generate license by license_id see example here
```
https://github.com/GluuFederation/oxd/blob/version_3.1.1/oxd-server/src/main/java/org/xdi/oxd/server/license/LicenseFileUpdateService.java#L83-83
```

It's up to application how often to regenerate license as well as how often to validate it. General recommendation is to validate license at least one time per 24h. We can provide kind of `offline_license` file which would mean that we generate license and give to customer. Customer put it in right place and application reads it and validates it (in this way we can use it without internet). However the beautify of online approach is that we can put different restrictions or extend license on the fly.

**Comment by iromli** [comment]
@afroDC as mentioned by @yuriyz,

> For license generation and validation it's required to have:
>
>    license_id
>    public_key
>    public_password
>    license_password

Using the `license_id`, we can obtain signed license (contains data of license; encoded), which can be decoded using offline `oxlicense-validator` (requires `public_key`, `public_password`, and `license_password`). We can verify the validity (and other things) of license based on decoded data from signed license.

**Comment by iromli** [comment]
As we're using Python app, we can install http://ox.gluu.org/maven/org/xdi/oxlicense-validator/3.1.1.Final/oxlicense-validator-3.1.1.Final-jar-with-dependencies.jar and execute the program using subprocess.

**Comment by iromli** [comment]
> It's up to application how often to regenerate license as well as how often to validate it. General recommendation is to validate license at least one time per 24h.

Cluster manager probably can use automatic scheduled task if needed. Or simply provide UI to regenerate and validate license updates whenever existing license is expired.

> We can provide kind of offline_license file which would mean that we generate license and give to customer. Customer put it in right place and application reads it and validates it (in this way 
> we can use it without internet). However the beautify of online approach is that we can put 
> different restrictions or extend license on the fly.

I'm agree with @yuriyz about using online approach.

**Comment by afroDC** [comment]
@iromli we should also implement a license timer of some sort with a notification of expiration message for the user to see.

**Comment by iromli** [comment]
@afroDC did you mean something like a reminder when license will be expired?

**Comment by iromli** [comment]
@afroDC is this reminder should be displayed on all pages?

**Comment by afroDC** [comment]
Yes

**Comment by iromli** [comment]
@afroDC @yuriyz currently there are 2 license products ("oxd" for oxD installation and "de" for old gluu-engine). Do we need a new product name or just use existing name?

**Comment by yuriyz** [comment]
for now use de temporary please, there is no any logic behind it. And please add ticket on me to add new name which u wish to use.

**Comment by afroDC** [comment]
https://github.com/GluuFederation/cluster-mgr/issues/86

Connecting these issues

**Comment by iromli** [comment]
Implemented in `2.1` branch.
